app/gen-server/ApiServer.ts

```diff
@@ -627,6 +636,147 @@ export class ApiServer {
+          hasValidKey: true
+        };
+
+        return sendOkReply(req, res, resp);
+      }));
+
+      // GET /service-accounts/
+      // Reads all service accounts attached to the user making the api call.
+      this._app.get('/api/service-accounts', expressWrap(async (req, res) => {
+        const userId = getAuthorizedUserId(req);
+        const data = await this._dbManager.getOwnedServiceAccounts(userId);
+        const resp: Array<Partial<SATypes.ServiceAccountApiResponse>> = data.map(serviceAccount => {
+          const hasValidKey = serviceAccount.serviceUser.apiKey !== null;
+          return {
+            id: serviceAccount.id,
+            label: serviceAccount.label,
+            description: serviceAccount.description,
+            expiresAt: serviceAccount.expiresAt.toISOString(),
+            hasValidKey
+          };
+        });
+        return sendOkReply(req, res, resp);
+      }));
+
+      // GET /service-accounts/:said
+      // Reads one particular service account of the user making the api call.
+      this._app.get('/api/service-accounts/:said', expressWrap(async (req, res) => {
+        const userId = getAuthorizedUserId(req);
+        const serviceAccountId = parseInt(req.params.said);
+        const serviceAccount = await this._dbManager.getServiceAccount(serviceAccountId);
+        this._dbManager.assertServiceAccountExistingAndOwned(serviceAccount, userId);
+        const hasValidKey = serviceAccount.serviceUser.apiKey !== null;
+        const resp: Partial<SATypes.ServiceAccountApiResponse> = {
+          id: serviceAccount.id,
+          label: serviceAccount.label,
+          description: serviceAccount.description,
+          expiresAt: serviceAccount.expiresAt.toISOString(),
+          hasValidKey
+        };
+        return sendOkReply(req, res, resp);
+      }));
+
+      // PATCH /service-accounts/:said
+      // Modifies one particular service account of the user making the api call.
+      this._app.patch('/api/service-accounts/:said', validateStrict(PatchServiceAccount), expressWrap(
+        async (req, res) => {
+          const userId = getAuthorizedUserId(req);
+          const serviceAccountId = parseInt(req.params.said);
+          const payload = req.body as SATypes.PatchServiceAccount;
+          const updateProps = {
+            ...( payload.label ? { label: payload.label } : {} ),
+            ...( payload.description ? { description: payload.description } : {} ),
+            ...( payload.expiresAt ? { expiresAt: new Date(payload.expiresAt) } : {} ),
+            expiresAt: payload.expiresAt !== undefined ? new Date(payload.expiresAt) : undefined,
+          };
+
+          const resp = await this._dbManager.updateServiceAccount(
+            serviceAccountId, updateProps, { expectedOwnerId: userId }
+          );
+          if (!resp) {
+            throw new ApiError(`No such service account as "${serviceAccountId}"`, 404);
+          }
```

Comment:
This should be unnecessary, `updateServiceAccount` normally raises an error if not found.